src/main/java/io/vertx/core/file/impl/AsyncFileImpl.java

```diff
@@ -459,16 +459,19 @@ private void checkContext() {
   }
 
   private void doClose(Handler<AsyncResult<Void>> handler) {
-    Future<Void> res = Future.future();
-    try {
-      ch.close();
-      res.complete(null);
-    } catch (IOException e) {
-      res.fail(e);
-    }
-    if (handler != null) {
-      vertx.runOnContext(v -> handler.handle(res));
-    }
+    ContextImpl handlerContext = vertx.getOrCreateContext();
+    context.executeBlocking((Future<Void> res) -> {
```

Comment:
execute on `handlerContext` directly and pass the `handler` as `resultHandler` of `executeBlocking`